File: go/internal/tasks/cloudtasks_test.go
package tasks

import (
	"context"
	"errors"
	"strings"
	"testing"
	"time"

	taskspb "cloud.google.com/go/cloudtasks/apiv2/cloudtaskspb"
	gax "github.com/googleapis/gax-go/v2"

	"automation-agent/internal/ingest"
)

// fakeSubmitter records the last CreateTask request and returns a configurable error.
type fakeSubmitter struct {
	req *taskspb.CreateTaskRequest
	err error
}

func (f *fakeSubmitter) CreateTask(_ context.Context, req *taskspb.CreateTaskRequest, _ ...gax.CallOption) (*taskspb.Task, error) {
	f.req = req
	if f.err != nil {
		return nil, f.err
	}
	return req.Task, nil
}

// newTestCloudTasks builds a CloudTasks over a fake submitter with a fixed clock, so task
// building is exercised without a live gRPC client.
func newTestCloudTasks(f *fakeSubmitter, token string) *CloudTasks {
	return &CloudTasks{
		client:      f,
		queuePath:   "projects/p/locations/l/queues/q",
		dispatchURL: "https://svc.run.app/internal/dispatch",
		token:       token,
		now:         func() time.Time { return time.Unix(1_700_000_000, 0).UTC() },
	}
}

func httpReq(t *testing.T, task *taskspb.Task) *taskspb.HttpRequest {
	t.Helper()
	hr, ok := task.MessageType.(*taskspb.Task_HttpRequest)
	if !ok {
		t.Fatalf("task message type = %T, want HttpRequest", task.MessageType)
	}
	return hr.HttpRequest
}

// A plain enqueue builds a POST task targeting /internal/dispatch, carrying the encoded
// envelope as the body and the INTERNAL_TOKEN as a Bearer header, with no name/schedule.
func TestCloudTasksBuildsTask(t *testing.T) {
	f := &fakeSubmitter{}
	ct := newTestCloudTasks(f, "sekret")
	env := ingest.New(ingest.KindCI, "webhook:/github", []byte(`{"action":"completed"}`), time.Unix(0, 0).UTC())

	if err := ct.Enqueue(context.Background(), env); err != nil {
		t.Fatalf("Enqueue: %v", err)
	}
	if f.req.Parent != "projects/p/locations/l/queues/q" {
		t.Errorf("parent = %q", f.req.Parent)
	}
	hr := httpReq(t, f.req.Task)
	if hr.HttpMethod != taskspb.HttpMethod_POST {
		t.Errorf("method = %v, want POST", hr.HttpMethod)
	}
	if hr.Url != "https://svc.run.app/internal/dispatch" {
		t.Errorf("url = %q", hr.Url)
	}
	if hr.Headers["Authorization"] != "Bearer sekret" {
		t.Errorf("auth header = %q", hr.Headers["Authorization"])
	}
	if hr.Headers["Content-Type"] != "application/json" {
		t.Errorf("content-type = %q", hr.Headers["Content-Type"])
	}
	// The body is the exact wire codec output, and decodes back to the envelope.
	want, _ := ingest.Encode(env)
	if string(hr.Body) != string(want) {
		t.Errorf("body = %s, want %s", hr.Body, want)
	}
	out, err := ingest.Decode(hr.Body)
	if err != nil || out.Kind != ingest.KindCI {
		t.Errorf("body did not round-trip: %v / %q", err, out.Kind)
	}
	if f.req.Task.Name != "" {
		t.Errorf("unexpected task name %q (no dedup requested)", f.req.Task.Name)
	}
	if f.req.Task.ScheduleTime != nil {
		t.Errorf("unexpected schedule time (no delay requested)")
	}
}

// The optional dedup name and schedule delay are carried onto the built task.
func TestCloudTasksHonorsNameAndDelay(t *testing.T) {
	f := &fakeSubmitter{}
	ct := newTestCloudTasks(f, "")
	env := ingest.New(ingest.KindCoverage, "webhook:/coverage", []byte("{}"), time.Unix(0, 0))

	if err := ct.Enqueue(context.Background(), env, WithName("pr-42"), WithDelay(30*time.Second)); err != nil {
		t.Fatalf("Enqueue: %v", err)
	}
	if got, want := f.req.Task.Name, "projects/p/locations/l/queues/q/tasks/pr-42"; got != want {
		t.Errorf("name = %q, want %q", got, want)
	}
	if f.req.Task.ScheduleTime == nil {
		t.Fatal("schedule time not set")
	}
	if got := f.req.Task.ScheduleTime.AsTime(); !got.Equal(time.Unix(1_700_000_030, 0).UTC()) {
		t.Errorf("schedule time = %v, want now+30s", got)
	}
	// With no token configured, no Authorization header is attached.
	if _, ok := httpReq(t, f.req.Task).Headers["Authorization"]; ok {
		t.Error("Authorization header set despite empty token")
	}
}

// An envelope whose encoded body exceeds the Cloud Tasks task-size limit is refused
// up front rather than failing opaquely at CreateTask (spec §9).
func TestCloudTasksRejectsOversizeEnvelope(t *testing.T) {
	f := &fakeSubmitter{}
	ct := newTestCloudTasks(f, "")
	big := ingest.New(ingest.KindLint, "s", make([]byte, MaxTaskBytes+1), time.Unix(0, 0))
	err := ct.Enqueue(context.Background(), big)
	if err == nil || !strings.Contains(err.Error(), "task limit") {
		t.Fatalf("err = %v, want task-limit error", err)
	}
	if f.req != nil {
		t.Error("oversize envelope should not reach CreateTask")
	}
}

// A CreateTask failure surfaces to the caller (which becomes a 500 → the webhook source
// retries, and the queue itself retries an /internal/dispatch failure).
func TestCloudTasksSurfacesSubmitError(t *testing.T) {
	f := &fakeSubmitter{err: errors.New("unavailable")}
	ct := newTestCloudTasks(f, "")
	err := ct.Enqueue(context.Background(), ingest.New(ingest.KindCI, "s", nil, time.Unix(0, 0)))
	if err == nil || !strings.Contains(err.Error(), "create task") {
		t.Fatalf("err = %v, want create-task error", err)
	}
}

// Close releases the underlying client, and is a no-op when none is set.
func TestCloudTasksClose(t *testing.T) {
	closed := false
	ct := &CloudTasks{closer: func() error { closed = true; return nil }}
	if err := ct.Close(); err != nil || !closed {
		t.Fatalf("Close: err=%v closed=%v", err, closed)
	}
	if err := (&CloudTasks{}).Close(); err != nil {
		t.Errorf("Close with no client = %v, want nil", err)
	}
}

File: javascript/src/tasks/cloudtasks.test.ts
// Tests for the Cloud Tasks execution transport (task-building, exercised against a fake
// submitter so no live gRPC client is needed).
import { protos } from '@google-cloud/tasks';
import { context, propagation, trace } from '@opentelemetry/api';
import { InMemorySpanExporter } from '@opentelemetry/sdk-trace-base';
import { afterEach, describe, expect, it } from 'vitest';

import { decode, encode, type Envelope, Kind, newEnvelope } from '../ingest/envelope';
import { TracesExporter, install } from '../obs/obs';
import { CloudTasks, MAX_TASK_BYTES, type Submitter } from './cloudtasks';

type ICreateTaskRequest = protos.google.cloud.tasks.v2.ICreateTaskRequest;

function env(kind: Kind = Kind.Lint, payload = 'x'): Envelope {
  return newEnvelope(kind, 'webhook:/lint', Buffer.from(payload), new Date(0));
}

/** Records the last CreateTaskRequest and returns a configurable error. */
class FakeSubmitter implements Submitter {
  request: ICreateTaskRequest | null = null;
  closed = false;
  constructor(private readonly err: Error | null = null) {}

  async createTask(request: ICreateTaskRequest): Promise<unknown> {
    this.request = request;
    if (this.err !== null) {
      throw this.err;
    }
    return [request.task];
  }

  async close(): Promise<void> {
    this.closed = true;
  }
}

/** Build a CloudTasks over a fake submitter with a fixed clock, so task building is exercised
 * without a live gRPC client. */
function newCT(f: FakeSubmitter, token: string): CloudTasks {
  return new CloudTasks({
    client: f,
    queuePath: 'projects/p/locations/l/queues/q',
    dispatchUrl: 'https://svc.run.app/internal/dispatch',
    token,
    deadlineMs: 30 * 60 * 1000,
    now: () => new Date(1_700_000_000 * 1000),
  });
}

describe('CloudTasks', () => {
  it('builds a POST task carrying the envelope and a Bearer token', async () => {
    const f = new FakeSubmitter();
    const ct = newCT(f, 'sekret');
    const e = newEnvelope(Kind.CI, 'webhook:/github', Buffer.from('{"action":"completed"}'), new Date(0));
    await ct.enqueue(e);

    const req = f.request!;
    expect(req.parent).toBe('projects/p/locations/l/queues/q');
    const hr = req.task!.httpRequest!;
    expect(hr.httpMethod).toBe('POST');
    expect(hr.url).toBe('https://svc.run.app/internal/dispatch');
    expect(hr.headers!.Authorization).toBe('Bearer sekret');
    expect(hr.headers!['Content-Type']).toBe('application/json');
    // The body is the exact wire codec output and decodes back to the envelope.
    expect((hr.body as Buffer).equals(encode(e))).toBe(true);
    expect(decode(hr.body as Buffer).kind).toBe(Kind.CI);
    // No dedup name / schedule requested.
    expect(req.task!.name == null).toBe(true);
    expect(req.task!.scheduleTime == null).toBe(true);
    // The dispatch deadline is set explicitly (so a long workflow is not cancelled at the
    // HTTP-target default of 10m and retried, duplicating side effects).
    expect(req.task!.dispatchDeadline?.seconds).toBe(30 * 60);
  });

  it('omits the dispatch deadline when unset (zero)', async () => {
    // With no deadline configured the task omits dispatchDeadline so the queue default applies —
    // production always supplies a config-validated value.
    const f = new FakeSubmitter();
    const ct = new CloudTasks({
      client: f,
      queuePath: 'projects/p/locations/l/queues/q',
      dispatchUrl: 'https://svc.run.app/internal/dispatch',
      token: '',
    });
    await ct.enqueue(env(Kind.CI, ''));
    expect(f.request!.task!.dispatchDeadline == null).toBe(true);
  });

  it('honors the dedup name and schedule delay', async () => {
    const f = new FakeSubmitter();
    const ct = newCT(f, '');
    await ct.enqueue(env(Kind.Coverage, '{}'), { name: 'pr-42', delayMs: 30_000 });

    expect(f.request!.task!.name).toBe('projects/p/locations/l/queues/q/tasks/pr-42');
    expect(f.request!.task!.scheduleTime?.seconds).toBe(1_700_000_030);
    // With no token configured, no Authorization header is attached.
    expect(f.request!.task!.httpRequest!.headers!.Authorization).toBeUndefined();
  });

  it('rejects an oversize envelope up front', async () => {
    // An envelope whose encoded body exceeds the Cloud Tasks task-size limit is refused before
    // create rather than failing opaquely at create time (spec §9).
    const f = new FakeSubmitter();
    const ct = newCT(f, '');
    const big = newEnvelope(Kind.Lint, 's', Buffer.alloc(MAX_TASK_BYTES + 1), new Date(0));
    await expect(ct.enqueue(big)).rejects.toThrow(/task limit/);
    expect(f.request).toBeNull(); // never reached createTask
  });

  it('surfaces a submit error to the caller', async () => {
    // A create failure surfaces (which becomes a 500 -> the webhook source retries, and the
    // queue retries an /internal/dispatch failure).
    const f = new FakeSubmitter(new Error('unavailable'));
    const ct = newCT(f, '');
    await expect(ct.enqueue(env(Kind.CI, ''))).rejects.toThrow(/create task/);
  });

  it('closes the underlying client', async () => {
    const f = new FakeSubmitter();
    await newCT(f, '').close();
    expect(f.closed).toBe(true);
  });
});

describe('CloudTasks trace propagation', () => {
  afterEach(() => {
    trace.disable();
    propagation.disable();
    context.disable();
  });

  it('injects a W3C traceparent header when tracing is enabled', async () => {
    const shutdown = install(new InMemorySpanExporter(), {
      exporter: TracesExporter.Console,
      serviceName: 'automation-agent',
    });
    try {
      const f = new FakeSubmitter();
      const ct = newCT(f, 'sekret');
      // Enqueue under an active span (the ingress request), so the task continues that trace.
      const span = trace.getTracer('t').startSpan('ingress');
      await context.with(trace.setSpan(context.active(), span), () => ct.enqueue(env(Kind.CI, '{}')));
      span.end();
      expect(f.request!.task!.httpRequest!.headers!.traceparent).toBeTruthy();
    } finally {
      await shutdown();
    }
  });

  it('adds no traceparent when tracing is disabled', async () => {
    // With tracing off (no provider registered), inject is a no-op — no traceparent leaks onto the
    // task.
    const f = new FakeSubmitter();
    await newCT(f, 'sekret').enqueue(env(Kind.CI, '{}'));
    expect(f.request!.task!.httpRequest!.headers!.traceparent).toBeUndefined();
  });
});
